libuavcan/include/uavcan/util/avl_tree.hpp

```diff
@@ -0,0 +1,389 @@
+                        temp = node;
+                        node = UAVCAN_NULLPTR;
+                    } else {
+                        *node = *temp;
+                    }
+
+                    len_--;
+                    deleteNode(temp);
+                } else {
+                    Node *minOfRight = node->right;
+
+                    while (minOfRight->left != UAVCAN_NULLPTR) {
+                        minOfRight = minOfRight->left;
+                    }
+
+                    node->data = minOfRight->data;
+                    node->right = remove_node(node->right, minOfRight->data);
+                }
+            } else {
+                Node* newHead = deleteFromList(node, data);
+                return newHead;
+            }
+        }
+
+        if (node == UAVCAN_NULLPTR) {
+            return node;
+        }
+
+        node->h = static_cast<int16_t>(maxOf(heightOf(node->left), heightOf(node->right)) + 1);
+
+        int16_t balance = balanceOf(node);
+
+        if (balance > 1 && balanceOf(node->left) >= 0) {
+            return rotateRight(node);
+        }
+
+        if (balance > 1 && balanceOf(node->left) < 0) {
+            node->left = rotateLeft(node->left);
+            return rotateRight(node);
+        }
+
+        if (balance < -1 && balanceOf(node->right) <= 0) {
+            return rotateLeft(node);
+        }
+
+        if (balance < -1 && balanceOf(node->right) > 0) {
+            node->right = rotateRight(node->right);
+            return rotateLeft(node);
+        }
+
+        return node;
+    }
+
+    bool linkedListContains(Node* head, const T* data) const {
+        Node* next = head;
+        while(next != UAVCAN_NULLPTR) {
+            if(next->data == data) { /* Memory address comparison */
+                return true;
+            }
+            next = next->equalKeys;
+        }
+        return false;
+    }
+
+public:
+    AvlTree(IPoolAllocator &allocator, std::size_t allocator_quota)
+        : root_(UAVCAN_NULLPTR)
+        , len_(0)
+        , allocator_(allocator, allocator_quota)
+    {}
+
+    virtual ~AvlTree() {
+        // delete leafs first
+        postOrderTraverseNodeCleanup(this->root_);
+    }
+
+    void remove_entry(T* data) {
+        root_ = remove_node(root_, data);
+    }
+
+    bool insert(T* data) {
+        Node* newNode = makeNode(data);
+        if(newNode == UAVCAN_NULLPTR) {
+            return false;
+        }
+
+        root_ = insert_node(root_, newNode);
+
+        return true;
+    }
+
+    size_t getSize() const {
+        return root_ == UAVCAN_NULLPTR ? 0 : len_;
+    }
+
+    void walkPostOrder(std::function<void(T*&)> forEach) {
+        postOrderTraverseRecursively(root_, forEach);
+    }
+
+
+    bool isEmpty() const {
+        return getSize() == 0;
+    }
+
+    T* max() const {
+        Node* n = root_;
+
+        if (n == UAVCAN_NULLPTR) {
+            return UAVCAN_NULLPTR;
+        }
+
+        for (;;) {
```

Comment:
MISRA does not permit unconditional loops outside of `main`. Can this be restructured?